hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java

```diff
@@ -0,0 +1,319 @@
+import com.hazelcast.sql.support.expressions.ExpressionType;
+import com.hazelcast.sql.support.expressions.ExpressionTypes;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.io.Serializable;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+import static com.hazelcast.jet.datamodel.Tuple2.tuple2;
+import static com.hazelcast.sql.SqlColumnType.BIGINT;
+import static com.hazelcast.sql.SqlColumnType.DATE;
+import static com.hazelcast.sql.SqlColumnType.DOUBLE;
+import static com.hazelcast.sql.SqlColumnType.INTEGER;
+import static com.hazelcast.sql.SqlColumnType.OBJECT;
+import static com.hazelcast.sql.SqlColumnType.TIME;
+import static com.hazelcast.sql.SqlColumnType.VARCHAR;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+/**
+ * Test for BETWEEN operator.
+ * <p>
+ * Hazelcast SQL engine has 13 types in its type system (also OBJECT and
+ * NULL, but they don't count in case of BETWEEN operator). BETWEEN
+ * operator has 3 arguments and the test should check 13*13*13 = 2197
+ * possible arguments type combination.
+ * <p>
+ * The BETWEEN operator has two possible modes:<ul>
+ *     <li>ASYMMETRIC, which is default mode. SQL engine converts "a BETWEEN b
+ *     AND c" to "a <= b AND a >= c"
+ *
+ *     <li>SYMMETRIC. It it can be used in cases where the user is not sure
+ *     that c >= b. SQL engine converts "a SYMMETRIC BETWEEN b AND c" to "(a <=
+ *     b AND a >= c) OR (a >= b AND a <= c)"
+ * </ul>
+ *
+ * <p>
+ * Types are represented by {@link ExpressionType} and its subclasses.
+ * Also, there are lists of values to write to an IMap for each type in
+ * {@link ExpressionType} subclasses.
+ *
+ * <p>
+ * Testing algorithm:<ul>
+ * <li>The test writes to IMap a collection of elements with the desired types.
+ * <li>It executes a query with an equivalent predicate not using BETWEEN
+ * (see above)
+ * <li>It intercepts the result or the exception
+ * <li>Finally, the test executes a query with BETWEEN operator and asserts
+ * the result or the exception is the same as for the equivalent query.
+ *
+ * @see #betweenAsymmetricPredicateTypeCheckTest()
+ * @see #betweenSymmetricPredicateTypeCheckTest()
+ */
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class BetweenOperatorIntegrationTest extends ExpressionTestSupport {
+
+    // TODO [viliam] remove OBJECT?
+    static final ExpressionType<?>[] TESTED_TYPES = ExpressionTypes.allExcept(ExpressionTypes.OBJECT);
```

Comment:
We probably should not ignore OBJECT, the behavior should be the same in this case too.